### PR TITLE
feat(evals): pass tool calls to code and LLM-as-a-judge evaluators

### PR DESCRIPTION
**Pull request by wochinge**
## Summary

- Code evaluators now receive the tool calls recorded on the target observation as `ctx.observation.toolCalls` (TypeScript) / `ctx.observation.tool_calls` (Python) — self-contained `{id, name, arguments, type, index}` objects. ClickHouse stores calls as name-less JSON strings with names in a parallel array (so filters need no JSON parsing); the new `zipObservationToolCalls` helper rebuilds named objects and parses the double-encoded `arguments`, and is shared by the worker runtime, the test-run preview, and the UI so all three agree ([LFE-10688](https://linear.app/langfuse/issue/LFE-10688/add-access-to-tool-calls-in-code-evals)).
- "Tool Calls" is now a mappable variable column for observation- and experiment-target evaluators, so LLM-as-a-judge prompts can reference called tools too, with JSONPath selectors like `$[*].name` ([LFE-10755](https://linear.app/langfuse/issue/LFE-10755/add-access-to-tools-called-in-llm-as-a-judge)).
- The unstable public API accepts and documents the `tool_calls` mapping source (Fern docs, zod enums, adapters). Validation now derives the allowed sources from the zod enums — the hand-rolled list had already drifted and 400'd the documented source (caught in review, covered by a regression test).
- Saving a new code-evaluator template version now rewrites the variable mapping of re-pointed evaluation rules to the canonical set (tRPC and public API), so rules created before this change receive tool calls the moment their template starts using them — instead of silently scoring with `[]` while the test run showed populated data.
- The code editor ships the updated contract (`ToolCall` type / `tool_calls` dataclass), hover docs, and starter examples; sources saved under previously shipped contract versions still strip cleanly.

## Linear

- [LFE-10688](https://linear.app/langfuse/issue/LFE-10688/add-access-to-tool-calls-in-code-evals)
- [LFE-10755](https://linear.app/langfuse/issue/LFE-10755/add-access-to-tools-called-in-llm-as-a-judge)

## Major Decisions

- **Tool calls flow through the variable-mapping machinery, not a payload side-channel.** An earlier iteration passed the observation's tool columns directly to `buildCodeEvalPayload`; routing through extraction keeps one code path for preview, test run, and runtime, and made the redundant `toolCallNames` payload field unnecessary (names live on each call).
- **Sync-on-template-save instead of a backfill migration.** Code-eval mappings are never user-authored, and a stale mapping only matters once the evaluator code uses a new variable — which requires saving the template. Both template-save flows already mutate the linked rules (re-pointing them to the new version), so they now also write the canonical mapping in the same transaction. This removes the migration entirely: no rolling-deploy re-drift window, and no recurring migration on future variable additions. The public API upgrade path previously 409ed for code evaluators whose stored snapshot was missing a newly added canonical variable; it now synthesizes instead.
- **batchIO tool columns are opt-in (`includeToolCallFields`).** `events_core` truncates I/O to 200 chars but copies `tool_calls` verbatim — fetching it unconditionally would ship the one uncapped payload column on every events-list page. Only the eval previews (single row, `truncated: false`) request it.

## Review Focus

- **Deploy ordering:** the Python Lambda runner must deploy before or with the worker — an old runner receiving the new payload drops `tool_calls`, and user code accessing it gets `AttributeError`. The Node runner passes the payload through verbatim and is unaffected.
- **Mapping sync on template save** (`router.ts` `createTemplate`, `evaluator-service.ts` upgrade path): re-pointed CODE rules adopt `getCodeEvalVariableMapping()` in the same transaction. Covered by a real-Postgres tRPC servertest (stale rule → new version → canonical mapping) and a service unit test for the public API path; LLM-judge mappings are untouched.
- Known follow-ups deliberately left out: parsing `rawToolCalls` with `toolCallForEvalSchema` instead of a cast, mentioning tool calls in the `PAYLOAD_TOO_LARGE` message, and two lockstep tests (`convertCallsToArrays` ↔ `zipObservationToolCalls` round-trip; Python contract verbatim-in-runner).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds tool call data to both code and LLM-as-a-judge evaluators. A new `zipObservationToolCalls` helper rebuilds named tool-call objects from ClickHouse's parallel-array storage layout, and is shared by the worker runtime, test-run preview, and UI so all three surfaces agree. A data migration resets stored code-eval variable mappings to the canonical set so pre-existing rules receive tool calls automatically.

- `zipObservationToolCalls` is added to `observationForEval.ts` and threaded through `extractObservationVariables`, `usePreviewData`, and `code-eval-test-run-card` to unify how tool calls are surfaced.
- The unstable public API adds `tool_calls` as a valid mapping source for both `observation` and `experiment` targets; `SUPPORTED_MAPPING_SOURCES_BY_TARGET` is now derived from the zod enums instead of a hand-rolled list.
- A SQL migration overwrites `variable_mapping` for all `CODE`-type `job_configurations` with the new canonical set (including `toolCalls`), and the Python Lambda handler gains `tool_calls: list[Any]` on `ObservationContext`.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

The PR is safe to merge with care paid to Python Lambda deploy ordering — new evaluators accessing `ctx.observation.tool_calls` will fail if an old runner is still serving requests when the worker deploys the new payload.

The core data path is well-tested: `zipObservationToolCalls` has dedicated unit tests covering empty arrays, malformed entries, and the already-parsed case; the extraction, dispatch, and preview paths agree on the same transformation. The `deepParseJson` call over the already-zipped `ToolCallForEval[]` is intentional and harmless for typical argument values, but argument strings that are valid JSON literals would be silently coerced. The migration is lossless and idempotent; the narrow rolling-deploy window during which new rules could store the stale mapping is the main operational risk.

`packages/shared/src/server/evals/extractObservationVariables.ts` (double-parse path) and `packages/shared/prisma/migrations/20260709120000_reset_code_eval_variable_mappings/migration.sql` (rolling-deploy window for new rules created by old pods).
</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant CH as ClickHouse
    participant EVT as events.ts
    participant EXT as extractObservationVariables
    participant ZIP as zipObservationToolCalls
    participant BLD as buildCodeEvalPayload
    participant WRK as Worker dispatch
    participant PY as Python Lambda
    participant PRV as usePreviewData

    CH->>EVT: "raw parallel arrays (includeToolCallFields=true)"
    EVT->>EXT: "ObservationForEval {tool_calls[], tool_call_names[]}"
    EXT->>ZIP: zipObservationToolCalls(obs)
    ZIP-->>EXT: "ToolCallForEval[] {id,name,arguments,type,index}"
    EXT-->>BLD: "ExtractedVariable {var:toolCalls, value:[...]}"
    BLD-->>WRK: CodeEvalPayload observation.toolCalls
    WRK->>PY: JSON payload to ctx.observation.tool_calls

    CH->>EVT: raw parallel arrays (preview)
    EVT->>PRV: "record {toolCalls:string[], toolCallNames:string[]}"
    PRV->>ZIP: zipObservationToolCalls(record)
    ZIP-->>PRV: ToolCallForEval[]
    PRV-->>PRV: deepParseJson(toolCalls) for display
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant CH as ClickHouse
    participant EVT as events.ts
    participant EXT as extractObservationVariables
    participant ZIP as zipObservationToolCalls
    participant BLD as buildCodeEvalPayload
    participant WRK as Worker dispatch
    participant PY as Python Lambda
    participant PRV as usePreviewData

    CH->>EVT: "raw parallel arrays (includeToolCallFields=true)"
    EVT->>EXT: "ObservationForEval {tool_calls[], tool_call_names[]}"
    EXT->>ZIP: zipObservationToolCalls(obs)
    ZIP-->>EXT: "ToolCallForEval[] {id,name,arguments,type,index}"
    EXT-->>BLD: "ExtractedVariable {var:toolCalls, value:[...]}"
    BLD-->>WRK: CodeEvalPayload observation.toolCalls
    WRK->>PY: JSON payload to ctx.observation.tool_calls

    CH->>EVT: raw parallel arrays (preview)
    EVT->>PRV: "record {toolCalls:string[], toolCallNames:string[]}"
    PRV->>ZIP: zipObservationToolCalls(record)
    ZIP-->>PRV: ToolCallForEval[]
    PRV-->>PRV: deepParseJson(toolCalls) for display
```

</a>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 3 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 3
web/src/features/evals/server/unstable-public-api/adapters.ts:96
**Defensive `tool_calls` key may silently absorb a future typo**

`INTERNAL_MAPPING_COLUMN_TO_PUBLIC_SOURCE` contains both `toolCalls` and `tool_calls` mapping to `"tool_calls"`. The canonical `selectedColumnId` stored in variable mappings is always `"toolCalls"` (camelCase), so the `tool_calls` (snake_case) key is never actually hit for tool calls — unlike `expectedOutput`/`experimentItemExpectedOutput` which exist because both forms appear in real stored data. If a future write path accidentally stores `"tool_calls"` as the column ID, this silent fallback would mask the inconsistency rather than surfacing it at the `InternalServerError` boundary in `toApiMappings`. A comment clarifying the intent (or removing the entry) would reduce confusion.

### Issue 2 of 3
packages/shared/src/server/evals/extractObservationVariables.ts:42-50
**`deepParseJson` applied to already-parsed `ToolCallForEval[]`**

`resolveFieldValue("tool_calls")` returns `ToolCallForEval[]` in which `arguments` is already a parsed object (not a JSON string). The subsequent `deepParseJson(fieldValue)` call then recursively walks the array and tries to parse any string values it encounters as JSON. For the typical case this is a no-op, but string argument values that happen to be valid JSON literals — e.g. `"42"`, `"true"`, or `"null"` — would be silently coerced to their parsed equivalents (`42`, `true`, `null`), changing the type that evaluator code receives. The preview in `code-eval-test-run-card.tsx` applies the same transformation intentionally for parity, but it's worth ensuring the test suite covers these edge cases, since the existing tests only verify non-parseable string arguments.

### Issue 3 of 3
packages/shared/src/server/evals/codeEvalExecution.ts:317-323
**Evaluators without `toolCalls` in stored mapping silently receive `[]`**

The comment above the block accurately describes the intent, but the current code path only protects evaluators whose stored `variable_mapping` is missing the `toolCalls` entry. After the migration runs, all existing rules are patched. However, any rule created by an old web pod **during** the rolling deploy window (after the migration but before the old pods are replaced) would still store the stale mapping and would receive `toolCalls: []` indefinitely until an operator manually updates the rule — there is no automatic self-correction unless the rule is re-saved. If a user's evaluator code reads `ctx.observation.tool_calls` and branches on it, it will silently produce wrong scores on those rules without any log evidence until the rule is updated.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(evals): address review findings for ..."](https://github.com/langfuse/langfuse/commit/64f2164c9a14d602bee0cace4743f4112042e18b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=42969281)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: packages/shared/src/features/evals/observationForEval.ts [hunk 1]
```diff
@@ -74,6 +75,51 @@ export const observationForEvalSchema = z.object({
...
+  name: z.string(),
+  arguments: z.unknown(),
+  type: z.string(),
+  index: z.number(),
+});
+
+export type ToolCallForEval = z.infer<typeof toolCallForEvalSchema>;
+
+const parseIfString = (value: unknown): unknown =>
+  typeof value === "string" ? parseJsonPrioritised(value) : value;
```
Comment (wochinge) [inline packages/shared/src/features/evals/observationForEval.ts]: we have similar helpers like this in the codebase for 3 times. Let's deduplicate

File: web/src/features/evals/server/unstable-public-api/adapters.ts [hunk 2]
```diff
@@ -91,6 +92,8 @@ const INTERNAL_MAPPING_COLUMN_TO_PUBLIC_SOURCE: Record<
   input: "input",
   output: "output",
   metadata: "metadata",
+  toolCalls: "tool_calls",
+  tool_calls: "tool_calls",
```
[the remark this answers is not in the document]
Comment (wochinge) [reply web/src/features/evals/server/unstable-public-api/adapters.ts]: Removed the snake_case `tool_calls` entry in 7cfbf5e1a. Tool calls are new with this PR, so unlike `expected_output` no legacy snake_case rows can exist — an accidental `tool_calls` write now surfaces at the corrupted-mapping `InternalServerError` boundary instead of being absorbed. Left a comment in the map explaining why only camelCase is listed.

File: packages/shared/src/server/evals/extractObservationVariables.ts [hunk 3]
```diff
@@ -24,17 +26,26 @@ export function extractObservationVariables(
...
     if (internal && observation[internal] !== undefined) {
+      const fieldValue = resolveFieldValue(internal);
       try {
-        parsedFields.set(fieldId, deepParseJson(observation[internal]));
+        parsedFields.set(fieldId, deepParseJson(fieldValue));
       } catch {
-        parsedFields.set(fieldId, observation[internal]);
+        parsedFields.set(fieldId, fieldValue);
       }
     }
```
[the remark this answers is not in the document]
Comment (wochinge) [reply packages/shared/src/server/evals/extractObservationVariables.ts:58]: Good instinct, but the coercion doesn't materialize for this structure: `deepParseJson`'s depth guard (default `maxDepth: 3` — array → call → arguments → leaf) stops before argument values, and bare number literals are additionally protected by the `isJsonNumberLiteral` guard (issue #6628). So `"42"`, `"true"`, and serialized-object string values inside `arguments` all pass through unchanged. Added a test in 7cfbf5e1a pinning exactly that, so any future change to the depth default breaks loudly.

File: packages/shared/src/features/evals/observationForEval.ts (hunk 1 above)
Comment (wochinge) [reply packages/shared/src/features/evals/observationForEval.ts]: Deduplicated in 7cfbf5e1a: added `parseJsonIfString` to `packages/shared/src/utils/json.ts` (precision-preserving via `parseJsonPrioritised`) and replaced all four copies — `observationForEval.ts`, `chatml/helpers.ts`, `extractToolsBackend.ts`, and `web/src/utils/chatml/extractTools.ts`. Note the ingestion/chatml sites move from plain `JSON.parse` to precision-preserving parsing: big integers now survive as strings instead of silently losing precision, matching the `parseJsonPrioritised` ingestion convention. Full extractToolsBackend + chatml suites pass.

File: web/src/features/events/server/eventsRouter.ts [hunk 4]
```diff
@@ -81,6 +81,7 @@ export const BatchIOInput = zodSchema.object({
   minStartTime: zodSchema.date(),
   maxStartTime: zodSchema.date(),
   truncated: zodSchema.boolean().optional(), // Defaults to true for performance
+  includeToolCalls: zodSchema.boolean().optional(), // Defaults to false; tool-call arrays can be large
```
[the remark this answers is not in the document]
Comment (wochinge) [reply web/src/features/events/server/eventsRouter.ts:86]: Fixed in 7cfbf5e1a: `RunEvaluationDialog` now requests `includeToolCalls: true` on `events.batchIO`, and `renderPromptPreviewFromObservation` zips both source shapes (events batchIO and legacy `observations.byId`) via a new shared `zipToolCallsFromRecord` — the same helper `usePreviewData` now uses — so `{{toolCalls}}` previews the named objects the evaluator receives.

File: fern/apis/server/definition/unstable/commons.yml [hunk 5]
```diff
@@ -58,19 +58,21 @@ types:
       Use these values when mapping evaluator prompt variables to live data.
 
       Target-specific rules:
-      - `target=observation` supports `input`, `output`, and `metadata`
-      - `target=experiment` supports `input`, `output`, `metadata`, `expected_output`, and `experiment_item_metadata`
+      - `target=observation` supports `input`, `output`, `metadata`, and `tool_calls`
+      - `target=experiment` supports `input`, `output`, `metadata`, `tool_calls`, `expected_output`, and `experiment_item_metadata`
```
[the remark this answers is not in the document]
Comment (wochinge) [reply fern/apis/server/definition/unstable/commons.yml:62]: Fixed in 7cfbf5e1a — both quick-reference bullets now include `tool_calls`.

**Comment by wochinge** [comment]
@claude review

File: web/src/features/evals/hooks/usePreviewData.ts [hunk 6]
```diff
@@ -90,6 +94,15 @@ function normalizePreviewDataFields(
...
 
+// Both source records (events batchIO and legacy observation) carry the raw
+// storage shape; zip client-side so the preview matches what the evaluator
+// runtime produces.
+function getZippedToolCalls(
+  record: Record<string, unknown> | null | undefined,
+) {
+  return zipToolCallsFromRecord(record ?? {});
+}
+
```
Comment (wochinge) [inline web/src/features/evals/hooks/usePreviewData.ts]: pointless helper let's remove

Comment (wochinge) [reply web/src/features/evals/hooks/usePreviewData.ts]: Removed in 909815cd2 — `normalizePreviewDataFields` calls `zipToolCallsFromRecord` directly.

**Comment by wochinge** [comment]
@claude review

**Comment by wochinge** [comment]
@claude review

**Comment by wochinge** [comment]
@claude review
